### PR TITLE
CONSOLE-4263: Add initial Content Security Policy for Console web application

### PR DESCRIPTION
**Pull request by vojtechszocs**
## Summary

This PR adds initial [Content Security Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) implementation for Console web application.

> Content Security Policy ([CSP](https://developer.mozilla.org/en-US/docs/Glossary/CSP)) is an added layer of security that helps to detect and mitigate certain types of attacks, including Cross-Site Scripting ([XSS](https://developer.mozilla.org/en-US/docs/Glossary/Cross-site_scripting)) and data injection attacks. These attacks are used for everything from data theft, to site defacement, to malware distribution.

Initially, we will deploy CSP in **report-only mode** via [`Content-Security-Policy-Report-Only`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy-Report-Only) HTTP header.

CSP violations are handled explicitly via Console [`SecurityPolicyViolationEvent`](https://developer.mozilla.org/en-US/docs/Web/API/SecurityPolicyViolationEvent) handler, which does the following:
- log the CSP event to browser console
- send the relevant CSP report data to Telemetry service

With this PR, running Console and Bridge locally with [dynamic demo plugin](https://github.com/openshift/console/tree/master/dynamic-demo-plugin) enabled should not trigger any CSP violations.

```sh
./bin/bridge -plugins console-demo-plugin=http://localhost:9001/ -i18n-namespaces=plugin__console-demo-plugin
```

**Note** :memo: since we use `Content-Security-Policy-Report-Only` instead of `Content-Security-Policy` header, the browser may warn about Console CSP not containing [`report-to`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/report-to) (standard) or [`report-uri`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/report-uri) (deprecated) directives as shown in the following screenshot:

This warning is not an issue, since we use a custom CSP violation event handler. The "cannot report violations" part of this warning means the browser itself will not make any CSP related HTTP requests.

The error below the warning in the screenshot is due to [Console HTML index template](https://github.com/openshift/console/blob/master/frontend/public/index.html) containing an inline script tag used to set up `SERVER_FLAGS` and visual theme config. The proper way to address this error is to allow this script tag - either [generate a SHA hash representing its contents](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#integrity) or [generate a cryptographically secure random token](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#nonce) for the script.

## Follow-up work

1. Address the above mentioned CSP inline script tag error.

2. Testing this PR has revealed that webpack generated Console assets require the following CSP exceptions, these should be removed in the future:

- [`script-src 'unsafe-eval'`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src#unsafe_eval_expressions) triggered by unsafe JS code like `eval()`
- [`style-src 'unsafe-inline'`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/style-src#unsafe_inline_styles) triggered by inline `<style>` elements

cc @spadgett @jhadvig

### CODE REVIEW

**Review by TheRealJon — COMMENTED** [review]
Looks good so far to me!

File: pkg/server/server.go
Comment (jhadvig) [inline pkg/server/server.go:207]: Please add [status-code]([status-code](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP#status-code)) as well.

File: pkg/server/server.go
Comment (jhadvig) [inline pkg/server/server.go:770]: this block should be moved above the ExecuteTemplate method execution
+ 
check if the `s.K8sMode` is set `off-cluster` or `on-cluster`

**Review by jhadvig — CHANGES_REQUESTED** [review]

File: pkg/server/server.go
Comment (vojtechszocs) [inline pkg/server/server.go:780]: Note - it is possible for Console frontend to send CSP violation reports directly to the appropriate service on the cluster.

If we decide to do that, we don't need this handler.

File: pkg/server/server.go
Comment (spadgett) [inline pkg/server/server.go:694]: Great start.

I could see some of these directives breaking plugins. As we've talked about, we'll minimally need a mechanism for plugins to add values to `cspSources` and a plan to transition to enforcing.

File: pkg/server/server.go
Comment (spadgett) [inline pkg/server/server.go:683]: We should determine if there is a more precise way to determine if we're using a webpack dev server in development mode. There are other use cases I know of where people run an off-cluster console.

File: frontend/public/components/app.jsx
Comment (spadgett) [inline frontend/public/components/app.jsx:169]: Is the violation automatically logged to the JS console as well?

We should discuss if there is a way during development mode to make it really obvious a violation occurred without the dev needing to explicitly check the console. Also we should think about the plugin development use case. We want to make sure plugin developers know they have violations that will break their plugin when we turn on enforcement.

**Review by spadgett — COMMENTED** [review]
Thanks, @vojtechszocs! Happy to see progress here.

File: pkg/server/server.go
Comment (vojtechszocs) [reply pkg/server/server.go:683]: @spadgett @jhadvig Any ideas on how to do this detection?

My initial idea was to add a new flag that would indicate the need to allow localhost CSP sources, i.e. during Console local development.

File: pkg/server/server.go
Comment (jhadvig) [reply pkg/server/server.go:694]: @spadgett 
> As we've talked about, we'll minimally need a mechanism for plugins to add values to cspSources

Sounds like an API change for the ConsolePlugin CRD

> plan to transition to enforcing.

I feel like by the end of 4.18 we should have a clear picture on:
1. What the CSP which we want to enforce should look like
2. API change for the ConsolePlugin

File: pkg/server/server.go
Comment (jhadvig) [reply pkg/server/server.go:683]: right, I think that should be the path forward

File: frontend/public/components/app.jsx
Comment (jhadvig) [reply frontend/public/components/app.jsx:169]: good idea. Maybe having a similar warnings like we do have with admission webhook, but just in the DEV mode?

**Comment by spadgett** [comment]
It would be good to have something in CI to report violations. If the Cypress tests can run today without violations, we could turn this on now to avoid regressions.

**Comment by jhadvig** [comment]
>It would be good to have something in CI to report violations. If the Cypress tests can run today without violations, we could turn this on now to avoid regressions.

@spadgett I agree. We have a [story](https://issues.redhat.com/browse/CONSOLE-4266) which should handle the one violation, console is currently reporting. I would suggest to add the CI check as a AC to this story. WDYT?
ccing @vojtechszocs

File: frontend/public/components/app.jsx
Comment (jhadvig) [inline frontend/public/components/app.jsx:175]: @vojtechszocs based on this the reported violation will have following format:
`"Content-Security-Policy violation | <cspReportObject> | <pluginName>"`

I assume that `cspReportObject` should contain details about the violation itself.

@luciddreamz could you please chime in here if this is a proper format, or if there is a better one. Thank you.

Comment (vojtechszocs) [reply frontend/public/components/app.jsx:175]: @jhadvig @luciddreamz Hi, CSP violation data sent from Console to telemetry should have the following format.

```ts
{

  // --- Standard CSP violation data ---

  // The URI of the resource that was blocked because it violates a policy.
  // https://developer.mozilla.org/en-US/docs/Web/API/SecurityPolicyViolationEvent/blockedURI
  blockedURI: string,
  // The column number in the document or worker at which the violation occurred.
  // https://developer.mozilla.org/en-US/docs/Web/API/SecurityPolicyViolationEvent/columnNumber
  columnNumber: number | null,
  // Whether the user agent is configured to enforce or just report the policy violation.
  // https://developer.mozilla.org/en-US/docs/Web/API/SecurityPolicyViolationEvent/disposition
  disposition: "enforce" | "report",
  // The URI of the document or worker in which the violation occurred.
  // https://developer.mozilla.org/en-US/docs/Web/API/SecurityPolicyViolationEvent/documentURI
  documentURI: string,
  // The directive that was violated.
  // https://developer.mozilla.org/en-US/docs/Web/API/SecurityPolicyViolationEvent/effectiveDirective
  effectiveDirective: string,
  // The line number in the document or worker at which the violation occurred.
  // https://developer.mozilla.org/en-US/docs/Web/API/SecurityPolicyViolationEvent/lineNumber
  lineNumber: number | null,
  // The policy whose enforcement caused the violation.
  // https://developer.mozilla.org/en-US/docs/Web/API/SecurityPolicyViolationEvent/originalPolicy
  originalPolicy: string,
  // The URL for the referrer of the resources whose policy was violated, or null.
  // https://developer.mozilla.org/en-US/docs/Web/API/SecurityPolicyViolationEvent/referrer
  referrer: string | null,
  // A sample of the resource that caused the violation, usually the first 40 characters.
  // https://developer.mozilla.org/en-US/docs/Web/API/SecurityPolicyViolationEvent/sample
  sample: string | null,
  // If the violation occurred as a result of a script, this will be the URL of the script.
  // https://developer.mozilla.org/en-US/docs/Web/API/SecurityPolicyViolationEvent/sourceFile
  sourceFile: string | null,
  // HTTP status code of the document or worker in which the violation occurred.
  // https://developer.mozilla.org/en-US/docs/Web/API/SecurityPolicyViolationEvent/statusCode
  statusCode: number,

  // --- OCP Console specific data ---

  // Name of the Console dynamic plugin that caused the violation.
  // Null value means the violation seems to originate from the Console application itself.
  pluginName: string | null,

}
```

Comment (vojtechszocs) [reply frontend/public/components/app.jsx:175]: @jhadvig @luciddreamz Sample CSP violation data instance:

```js
{
  "blockedURI": "eval",
  "columnNumber": 8,
  "disposition": "report",
  "documentURI": "http://localhost:9000/dashboards",
  "effectiveDirective": "script-src",
  "lineNumber": 208990,
  "originalPolicy": "default-src 'self' http://localhost:8080 ws://localhost:8080; base-uri 'self' http://localhost:8080 ws://localhost:8080; img-src 'self' http://localhost:8080 ws://localhost:8080 data:; font-src 'self' http://localhost:8080 ws://localhost:8080 data:; script-src 'self' http://localhost:8080 ws://localhost:8080; style-src 'self' http://localhost:8080 ws://localhost:8080 'unsafe-inline'; frame-src 'none'; frame-ancestors 'none'; object-src 'none'",
  "referrer": "",
  "sample": "",
  "sourceFile": "http://localhost:9000/static/vendors~main-8601b679a80dba767713.js",
  "statusCode": 200,
  "pluginName": null,
}
```

It seems that Firefox uses empty strings instead of null values, but that can be sanitized as a follow-up.

**Comment by vojtechszocs** [comment]
This PR needs an update, putting it on hold for now.

/hold

**Comment by vojtechszocs** [comment]
PR updated, it's ready for review.

**Review by jhadvig — APPROVED** [review]
Thank you @vojtechszocs 👍  The initial PR looks really good.
Looking forward for follow up changes.

/approve
/lgtm

**Comment by jhadvig** [comment]
QE Approver:
/assign @yapei
Docs Approver:
/assign @opayne1
PX Approver:
/assign @reestr

**Comment by spadgett** [comment]
I'd suggest we explore adding Cypress tests now if there are no known violations. Maybe there is something simple we can do to catch any CSP violation like we do for other generic runtime errors.

It looks like there are some special considerations for Cypress, though:

https://docs.cypress.io/guides/guides/content-security-policy

**Comment by jhadvig** [comment]
@spadgett yes we discussed with @vojtechszocs that the CI implementation would be done as a followup, but probably we should first address the [one violation console is currently triggering](https://issues.redhat.com/browse/CONSOLE-4266).

File: frontend/packages/console-dynamic-plugin-sdk/README.md
Comment (opayne1) [inline frontend/packages/console-dynamic-plugin-sdk/README.md:180]:
```suggestion
(CSP) to detect and mitigate certain types of attacks. By default, the list of allowed
```

File: frontend/packages/console-dynamic-plugin-sdk/README.md
Comment (opayne1) [inline frontend/packages/console-dynamic-plugin-sdk/README.md:182]:
```suggestion
includes the document origin `'self'` and Console webpack dev server when running off-cluster.
```

File: frontend/packages/console-dynamic-plugin-sdk/README.md
Comment (opayne1) [inline frontend/packages/console-dynamic-plugin-sdk/README.md:184]:
```suggestion
All dynamic plugin assets _should_ be loaded with the `/api/plugins/<plugin-name>` Bridge endpoint which
```

**Review by opayne1 — COMMENTED** [review]
Just a couple docs nits but otherwise looks good to me! I'll go ahead and add the label.

/label docs-approved

**Comment by vojtechszocs** [comment]
Thanks @opayne1 for your review, PR updated.

File: pkg/server/server.go
Comment (yapei) [inline pkg/server/server.go:688]: sorry, what policy does it mean with an empty `data:` scheme?

Comment (yapei) [inline pkg/server/server.go:688]: sorry, what does an empty `data:` schema mean here?

File: frontend/public/components/app.jsx
Comment (yapei) [inline frontend/public/components/app.jsx:182]: when dynamic plugin triggers CSP violation, we can only see warning message `Content Security Policy violation detected`

``Content Security Policy violation seems to originate from plugin ${pluginName}`` doesn't print

Screenshot 2024-09-27 at 4 46 18 PM

**Comment by yapei** [comment]
Deployed our [console-demo-plugin ](https://raw.githubusercontent.com/openshift/console/refs/heads/master/dynamic-demo-plugin/oc-manifest.yaml) and tested, can see some CSP violation errors coming from
- `/dynamic-route-1`
- GET https://console.redhat.com/connections/cdn/analytics.js/v1/BnuS1RP39EmLQjP21ko67oDjhbl9zpNU/analytics.min.js request

File: pkg/server/server.go
Comment (vojtechszocs) [reply pkg/server/server.go:688]: @yapei This allows the web application to use [`data:` URLs](https://developer.mozilla.org/en-US/docs/Web/URI/Schemes/data) to embed certain content into the HTML markup.

In this case, we allow using `data:` URLs for [`img-src`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/img-src) (images and favicons) and [`font-src`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/font-src) (CSS fonts).

Using `data:` URLs is commonly used for web asset optimization. When building Console web application, the module bundler (webpack) may generate these `data:` URLs for any small-sized images and fonts.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/Sources#scheme-source

File: frontend/public/components/app.jsx
Comment (vojtechszocs) [reply frontend/public/components/app.jsx:182]: @yapei @jhadvig 

When detecting whether the CSP violation originates from a dynamic plugin or from Console application code itself, we look at `blockedURI` and `sourceFile` properties - if their values start with `{consoleBaseURL}/api/plugins/` prefix, it means that the associated resource is fetched via Console Bridge server's plugin asset endpoint which allows us to infer the name of the dynamic plugin.

This is why the message is phrased "seems to originate from plugin" instead of just "originates from plugin". Any CSP violations which are due to `{consoleBaseURL}/api/plugins/...` requests are associated with dynamic plugins, _but any dynamic plugin can also make requests to other (non-Console) endpoints_.

If a plugin makes requests to non-Console endpoints (regardless of whether the requested service is inside or outside the cluster) it will trigger CSP violations _which we have no way of associating with that dynamic plugin_. These kind of CSP violations will be logged, but the "seems to originate from plugin" will not apply to them.

**Each dynamic plugin team should test their own plugin on a cluster and ensure there are no CSP violations that may be traced back to their plugin.**

In case a plugin really needs to make requests to non-Console endpoints then it should explicitly list such endpoints via `ConsolePlugin` custom resource ([CONSOLE-4265](https://issues.redhat.com/browse/CONSOLE-4265)).

**Comment by vojtechszocs** [comment]
@yapei @jhadvig 

> Deployed our [console-demo-plugin ](https://raw.githubusercontent.com/openshift/console/refs/heads/master/dynamic-demo-plugin/oc-manifest.yaml) and tested, can see some CSP violation errors in browser console
> 
> ```
> Content-Security-Policy: (Report-Only policy) The page's settings would block a script (script-src-elem) at https://console.redhat.com/connections/cdn/next-integrations/actions/845/d41568b7f25714884231.js from being executed because it violates the following directive: "script-src 'self' 'unsafe-eval'"
> 
> Content-Security-Policy: (Report-Only policy) The page's settings would block the loading of a resource (connect-src) at https://console.redhat.com/connections/cdn/v1/projects/BnuS1RP39EmLQjP21ko67oDjhbl9zpNU/settings because it violates the following directive: "default-src 'self'" analytics.min.js:1:23752
> 
> Content-Security-Policy: (Report-Only policy) The page's settings would block the loading of a resource (connect-src) at https://api.segment.io/v1/m because it violates the following directive: "default-src 'self'"
> ```

These CSP violations are not due to dynamic demo plugin - these are non-Console endpoints:
```
https://console.redhat.com/...
https://api.segment.io/...
```

As mentioned in my [comment](https://github.com/openshift/console/pull/14156#discussion_r1784861531) above, any requests to non-Console endpoints will trigger CSP violations according to the current Console policy.

We should determine whether ^^ endpoints are to be allowed on plugin level or on Console application level.

File: pkg/server/server.go
Comment (jhadvig) [reply pkg/server/server.go:688]: yes, this should be a valid value 👍

**Comment by jhadvig** [comment]
Yes, we need to determine where these endpoints are being called from and add them as en exception.
If its on the Console application level, we should extend the `cspSources` variable, or refactor the variable into a separate source based on the policy type.
If its on the plugin level, we will use the [new API field](https://github.com/openshift/api/pull/2042).
Either way this can be done as follow up 👍

**Comment by vojtechszocs** [comment]
> ```
> https://console.redhat.com/...
> https://api.segment.io/...
> ```

These endpoints are used by the Console application, so we'll need a follow-up change to include them in `cspSources` when generating the Console CSP. In the meantime, we should treat these CSP violations as known issues that we plan to fix in the near future.

**Comment by yapei** [comment]
Thanks for detailed explanation @vojtechszocs 
/label qe-approved

**Pull request merged**
